kv: don't remove voter replicas in ChangeReplicasTrigger

nvb opened the pull request:
Informs #129796.

This commit is a natural continuation of 7f37eaa (from 5 years ago). The commit made it so that the replicateQueue would never issue a ChangeReplicas request that removed a voter replica. Instead, the replicate queue would always demote the voter replica to a learner replica first.

This commit makes it so that the ChangeReplicasTrigger code does not even know how to remove a voter replica, it only knows how to demote them. This prevents abuse and strengthens the guarantee that we will never perform such unsafe configuration changes.

Release note: None

nvb committed:
kv: don't remove voter replicas in ChangeReplicasTrigger

Informs #129796.

This commit is a natural continuation of 7f37eaa (from 5 years ago). The commit
made it so that the replicateQueue would never issue a ChangeReplicas request
that removed a voter replica. Instead, the replicate queue would always demote
the voter replica to a learner replica first.

This commit makes it so that the ChangeReplicasTrigger code does not even know
how to remove a voter replica, it only knows how to demote them. This prevents
abuse and strengthens the guarantee that we will never perform such unsafe
configuration changes.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/130660)
<!-- Reviewable:end -->

nvb committed:
kv: don't construct raftpb.ConfChange with voter removal

Informs #129796.

This commit makes it so that the confChangeImpl code will never create a
raftpb.ConfChange(,V2) with a direct voter removal. Instead, it will always
demote the voter replica to a learner replica first.

Release note: None

nvanbenschoten commented:
TFTR!

bors r+
